rs/crypto/internal/crypto_lib/bls12_381/vetkd/tests/tests.rs

```diff
@@ -1,60 +1,480 @@
-    let _k = tsk.decrypt(&ek, &dpk, did).unwrap();
+    // Check that if we introduce incorrect shares then combine_all will fail
 
-    let derived_key = tsk
-        .decrypt_and_hash(&ek, &dpk, did, 32, b"aes-256-gcm-siv")
-        .unwrap();
-    assert_eq!(derived_key.len(), 32);
+    let other_did = rng.gen::<[u8; 24]>();
+    assert_ne!(proto.did, other_did);
+    let node_info_wrong_did = proto.create_encrypted_key_shares(rng, Some(&other_did));
+
+    let node_eks_wrong_did = node_info_wrong_did
+        .iter()
+        .map(|(idx, _pk, eks)| (*idx, eks.clone()))
+        .collect::<Vec<_>>();
+
+    // With combine_all even if we provide sufficiently many valid shares
+    // if any one share is invalid then combination will fail
+    for rec_threshold in 2..nodes {
+        let mut shares = random_subset(rng, &node_eks, rec_threshold - 1);
+
+        // Avoid using a duplicate index for this test
+        let random_unused_idx = loop {
+            let idx = (rng.gen::<usize>() % node_eks_wrong_did.len()) as u32;
+
+            if !shares.iter().map(|x| x.0).any(|x| x == idx) {
+                break idx as usize;
+            }
+        };
+
+        shares.push(node_eks_wrong_did[random_unused_idx].clone());
+        shares.shuffle(rng);
+
+        let expected_error = if rec_threshold < threshold {
+            EncryptedKeyCombinationError::InsufficientShares
+        } else {
+            EncryptedKeyCombinationError::InvalidShares
+        };
+        assert_eq!(proto.combine_all(&shares), Err(expected_error));
+    }
+
+    // Check that duplicate node indexes are detected
+    for rec_threshold in 2..nodes {
+        let mut shares = random_subset(rng, &node_eks, rec_threshold - 1);
+
+        let random_duplicate_idx = loop {
+            let idx = (rng.gen::<usize>() % node_eks.len()) as u32;
+
+            if shares.iter().map(|x| x.0).any(|x| x == idx) {
+                break idx as usize;
+            }
+        };
+
+        shares.push(node_eks[random_duplicate_idx].clone());
+        shares.shuffle(rng);
+
+        let expected_error = if rec_threshold < threshold {
+            EncryptedKeyCombinationError::InsufficientShares
+        } else {
+            EncryptedKeyCombinationError::DuplicateNodeIndex
+        };
+        assert_eq!(proto.combine_all(&shares), Err(expected_error));
+    }
+
+    // With combine_valid_shares OTOH we detect and reject the invalid shares
+
+    for rec_threshold in threshold..nodes {
+        let mut shares = random_subset(rng, &node_info, rec_threshold);
+        shares.append(&mut random_subset(rng, &node_info_wrong_did, 4));
+        shares.shuffle(rng);
+
+        let combined = proto.combine_valid(&shares).expect("Combination worked");
+
+        let k = setup
+            .transport_sk
+            .decrypt(&combined, &proto.derived_pk, &proto.did)
+            .expect("Decryption failed");
+
+        assert_eq!(k, vetkey);
+    }
+
+    for rec_threshold in threshold..nodes {
```

Comment:
For all other `for rec_threshold...` sections there is a comment describing what is tested. It would be helpful if such a comment was also added here.

Reply:
I added a comment describing what is being tested here

Reply:
I added a comment describing what is being tested here